datafusion/core/src/physical_plan/joins/symmetric_hash_join.rs

```diff
@@ -0,0 +1,2640 @@
+            Arc::new(RepartitionExec::try_new(
+                left,
+                Partitioning::Hash(left_expr, partition_count),
+            )?),
+            Arc::new(RepartitionExec::try_new(
+                right,
+                Partitioning::Hash(right_expr, partition_count),
+            )?),
+            on,
+            filter,
+            join_type,
+            &null_equals_null,
+        )?;
+
+        let mut batches = vec![];
+        for i in 0..partition_count {
+            let stream = join.execute(i, context.clone())?;
+            let more_batches = common::collect(stream).await?;
+            batches.extend(
+                more_batches
+                    .into_iter()
+                    .filter(|b| b.num_rows() > 0)
+                    .collect::<Vec<_>>(),
+            );
+        }
+
+        Ok(batches)
+    }
+    #[allow(clippy::too_many_arguments)]
+    async fn partitioned_hash_join_with_filter(
+        left: Arc<dyn ExecutionPlan>,
+        right: Arc<dyn ExecutionPlan>,
+        on: JoinOn,
+        filter: JoinFilter,
+        join_type: &JoinType,
+        null_equals_null: bool,
+        context: Arc<TaskContext>,
+    ) -> Result<Vec<RecordBatch>> {
+        let partition_count = 4;
+
+        let (left_expr, right_expr) = on
+            .iter()
+            .map(|(l, r)| (Arc::new(l.clone()) as _, Arc::new(r.clone()) as _))
+            .unzip();
+
+        let join = HashJoinExec::try_new(
+            Arc::new(RepartitionExec::try_new(
+                left,
+                Partitioning::Hash(left_expr, partition_count),
+            )?),
+            Arc::new(RepartitionExec::try_new(
+                right,
+                Partitioning::Hash(right_expr, partition_count),
+            )?),
+            on,
+            Some(filter),
+            join_type,
+            PartitionMode::Partitioned,
+            &null_equals_null,
+        )?;
+
+        let mut batches = vec![];
+        for i in 0..partition_count {
+            let stream = join.execute(i, context.clone())?;
+            let more_batches = common::collect(stream).await?;
+            batches.extend(
+                more_batches
+                    .into_iter()
+                    .filter(|b| b.num_rows() > 0)
+                    .collect::<Vec<_>>(),
+            );
+        }
+
+        Ok(batches)
+    }
+
+    pub fn split_record_batches(
+        batch: &RecordBatch,
+        batch_size: usize,
+    ) -> Result<Vec<RecordBatch>> {
+        let row_num = batch.num_rows();
+        let number_of_batch = row_num / batch_size;
+        let mut sizes = vec![batch_size; number_of_batch];
+        sizes.push(row_num - (batch_size * number_of_batch));
+        let mut result = vec![];
+        for (i, size) in sizes.iter().enumerate() {
+            result.push(batch.slice(i * batch_size, *size));
+        }
+        Ok(result)
+    }
+
+    fn build_record_batch(columns: Vec<(&str, ArrayRef)>) -> Result<RecordBatch> {
```

Comment:
This looks very similar to https://docs.rs/arrow/33.0.0/arrow/record_batch/struct.RecordBatch.html#method.try_from_iter

Reply:
Yeah, it is. Resolved 😄